BUGFIX: Asset usage prevent orphaned asset usage entries in case of a rebase with conflicts

mhsdesign opened the pull request:
Some projections / catchup hooks attempt to optimise to make `WorkspaceWasRebased` a no-op under the idea that all changes in the user workspace are still existent as known previously.

This is mostly true except for the case of a force rebase. Here changes in the user workspace might be discarded. To signal that to the consumer instead of a new even we introduce `WorkspaceWasRebased::hasSkippedEvents()`.

The asset usage was adjusted to flush all entries on force rebase so the state is rebuild from scratch without any orphaned entries that would only be deleted on a discard at some point.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Add note why we dont handle `WorkspaceWasPublished`

mhsdesign committed:
FEATURE: Indicate if a rebase had conflicts

mhsdesign committed:
BUGFIX: Asset usage prevent orphaned asset usage entries in case of a rebase with conflicts

dlubitz commented on Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php:
```diff
@@ -73,6 +74,13 @@ public function onAfterEvent(EventInterface $eventInstance, EventEnvelope $event
             }
         }
 
+        if ($eventInstance instanceof WorkspaceWasRebased && $eventInstance->hadConflicts) {
+            // because we don't know which changes were discarded in a conflict, we discard all changes and will build up the index on succeeding calls (with the kept reapplied events)
+            $this->discardWorkspace($eventInstance->getWorkspaceName());
+            return;
+        }
```
This could also go into the match below.

What happens to the remaining events if the workspace was rebased sucessfully? Aren't they re-applied too? So do we really need to distinguish here?

dlubitz replied:
Ok, question was answered in weekly. 

But still this can go into the match below.

mhsdesign replied:
the match would be slightly bloated then and i prefer to have the special logic here explicitly?

Though this will also work:

```php
    // because we don't know which changes were discarded in a conflict, we discard all changes and will build up the index on succeeding calls (with the kept reapplied events)
    WorkspaceWasRebased::class => $eventInstance->hadConflicts && $this->discardWorkspace($eventInstance->getWorkspaceName())
```

actually not that bad ... will do

dlubitz commented on Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Event/WorkspaceWasPublished.php:
```diff
@@ -58,6 +59,11 @@ public static function fromArray(array $values): self
         );
     }
 
+    public function getWorkspaceName(): WorkspaceName
+    {
+        return $this->sourceWorkspaceName;
+    }
+
```
Ohhh,... hmm. This seems to be dangerous. In your case "workspaceName" is "sourceWorkspaceName" but in other cases its "targetWorkspaceName". That's why we didn't implelemt the interface here earlier.

Also, is that needed at any point?

mhsdesign replied:
> Also, is that needed at any point?

Actually not 100% but i stumbled upon this while reading the code and this will be not fun to work with. The reason why i think the returning the `sourceWorkspaceName` is correct is because it is the intention of the command. Actually the fact that the `targetWorkspaceName` is part of the command is NEVER evaluated anywhere and might as well be removed (not really ^^) but its just metadata.

dlubitz replied:
I'd prefer to remove the interface in this case.

mhsdesign replied:
Removed and extracted discussion into separate pr: https://github.com/neos/neos-development-collection/pull/5431

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceRebase/Event/WorkspaceWasRebased.php:
```diff
@@ -34,6 +35,10 @@ public function __construct(
+        /**
+         * Indicates if failing changes were discarded during a forced rebase {@see RebaseErrorHandlingStrategy::STRATEGY_FORCE} or if all events in the workspace were kept.
+         */
+        public bool $hadConflicts
```
as discussed in the daily we want to rename this to hasSkippedEvents

mhsdesign committed:
TASK: Rename to `hasSkippedEvents` and store originally conflicting sequence numbers in force-rebase event

Storing the skipped sequence numbers should ease debugging rather than having 'just' a boolean indicator. The public api does not need to know about that thus the getter.

mhsdesign committed:
TASK: Inline condition into match call

dlubitz reviewed (approved):
This looks awesome. Thank you

dlubitz commented:
@mhsdesign Can we merge this?
